add stats that can only be collected at runtime

zhaojuanmao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #51394 log newly added construction and runtime stats at randomly selected iterations
* **#51386 add stats that can only be collected at runtime**

add stats such as rebuilt bucket stats, unused parameter stats and performance stats to ddp logging data

1. gpu time stats are not collected for single process multiple devices and multi device modules in this diff, as that requires events are created and recorded on multiple devices
2. use at::cuda::event API for safer calls
3. events may not be created in autograd hook if hook is not triggered in user's codes, e.g., users runs in non-sync mode in some iterations. So we checked events are created or not before synchronizing, also skipped invalid results.
4. users may not set device upfront, so explicitly set proper device before creating events in our prepare_forward() and prepare_backward() calls

Differential Revision: [D26158645](https://our.internmc.facebook.com/intern/diff/D26158645/)

zhaojuanmao committed:
add stats that can only be collected at runtime

add stats such as rebuilt bucket stats, unused parameter stats and performance stats to ddp logging data

Differential Revision: [D26158645](https://our.internmc.facebook.com/intern/diff/D26158645/)

[ghstack-poisoned]

SciPioneer commented on torch/lib/c10d/reducer.cpp:
```diff
@@ -1498,16 +1526,18 @@ void Reducer::set_parameter_stats() {
 }
 
-void Reducer::set_bucket_stats() {
+std::string Reducer::get_bucket_stats() {
```
Why is this a string?

SciPioneer replied:
If the stats only mean "sizes", it will be better to rename this function as "get_bucket_sizes". Stats sound very vague.

zhaojuanmao replied:
because some logger only support string, but you are right, let me change it back to vector<int> in struct level, and log it as string for some logger

SciPioneer commented on torch/lib/c10d/reducer.cpp:
```diff
@@ -1498,16 +1526,18 @@ void Reducer::set_parameter_stats() {
     }
-    ddp_logging_data_->bucket_sizes += std::to_string(bucket_size);
-    ddp_logging_data_->bucket_sizes += ", ";
+    bucket_sizes += std::to_string(bucket_size);
```
String concatenation in a loop is inefficient, especially if there are many buckets.

Better to construct a vector with a reserved capacity by `reserve`, and then casts it into a string when it needs to be displayed. I think we can use `folly::join` for more efficient string concatenation.

zhaojuanmao replied:
comment as above, will change it as vector<int>. meanwhile pytorch does not support to use folly though

SciPioneer replied:
You can also use `std::ostringstream` for casting the vector to a string for efficiency.

zhaojuanmao committed:
Update on "add stats that can only be collected at runtime"


add stats such as rebuilt bucket stats, unused parameter stats and performance stats to ddp logging data

Differential Revision: [D26158645](https://our.internmc.facebook.com/intern/diff/D26158645/)

[ghstack-poisoned]

zhaojuanmao commented:
right now all the timers work for CPU only, will add GPU timers

zhaojuanmao committed:
Update on "add stats that can only be collected at runtime"


add stats such as rebuilt bucket stats, unused parameter stats and performance stats to ddp logging data

Differential Revision: [D26158645](https://our.internmc.facebook.com/intern/diff/D26158645/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "add stats that can only be collected at runtime"


add stats such as rebuilt bucket stats, unused parameter stats and performance stats to ddp logging data

Differential Revision: [D26158645](https://our.internmc.facebook.com/intern/diff/D26158645/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "add stats that can only be collected at runtime"


add stats such as rebuilt bucket stats, unused parameter stats and performance stats to ddp logging data

Differential Revision: [D26158645](https://our.internmc.facebook.com/intern/diff/D26158645/)

[ghstack-poisoned]

rohan-varma commented on torch/lib/c10d/logger.cpp:
```diff
@@ -85,6 +87,101 @@ void Logger::set_construction_logging_data(
+  }
+  // rebuilt_bucket_sizes will not change once buckets are rebuilt,
+  // so it only needs to set once during whole training loop.
+  if (reducer_->num_iterations_ == 2) {
```
This seems like it could be bugprone since we are 1) hardcoding the iteration number and 2) Rebuild bucket logic can change in DDP, and this could start to silently be incorrect. 

Could we just keep `ddp_logging_data_->has_rebuilt_buckets ` in sync with `reducer_->has_rebuilt_bucket_;` and set the bucket sizes whenever the latter changes?

zhaojuanmao replied:
good idea!

rohan-varma commented on torch/lib/c10d/logger.cpp:
```diff
@@ -85,6 +87,101 @@ void Logger::set_construction_logging_data(
+
+  // set_runtime_stats is called at beginning of forward call,
+  // when it is cheap to synchronize the cuda events of previous iteration,
+  // as it is guaranteed that all cuda operations are done in previous
```
By "done" do we mean submitted + synchronized with the stream, or actually finished execution? I think we only guarantee the former AFAIK. Does this break the assumption here?

zhaojuanmao replied:
should not break the assumption. it assumes mostly users called optimizer.step() and all executions are done here. so it is cheap or no-op to call synchronize() here.  

even executions are not completely finished at this point in some rare cases. synchronize() will block and make sure time measurement is correct here. 

let me refactor the comment here as well

rohan-varma commented on torch/lib/c10d/logger.cpp:
```diff
@@ -85,6 +87,101 @@ void Logger::set_construction_logging_data(
+  // as it is guaranteed that all cuda operations are done in previous
+  // iteration.
+  if (reducer_->replicas_[0][0].is_cuda()) {
+#ifdef USE_CUDA
```
This compile-time flag should always be true in this `if` statement right? Otherwise how could `is_cuda()` be true?

rohan-varma replied:
Ah nevermind, one is compile time and the other is runtime

rohan-varma commented on torch/lib/c10d/logger.cpp:
```diff
@@ -85,6 +87,101 @@ void Logger::set_construction_logging_data(
+    int64_t& avg_time,
+    cudaEvent_t gpu_start,
+    cudaEvent_t gpu_end) {
+  long num_iters = reducer_->num_iterations_;
```
Should we add a TORCH_CHECK(num_iters > 0) here to be sure this is not called with num_iters = 0?

zhaojuanmao replied:
sounds good. set_runtime_stats() returned when num_iters <=0, but it is good to check here as well

rohan-varma commented on torch/lib/c10d/logger.cpp:
```diff
@@ -85,6 +87,101 @@ void Logger::set_construction_logging_data(
+  float milliseconds = 0.0;
+  cudaEventElapsedTime(&milliseconds, gpu_start, gpu_end);
+  avg_time =
+      (int(milliseconds * 1000000) + avg_time * (num_iters - 1)) / num_iters;
```
Nit: Use constant instead of magic number

zhaojuanmao replied:
sounds good

rohan-varma commented on torch/lib/c10d/logger.cpp:
```diff
@@ -85,6 +87,101 @@ void Logger::set_construction_logging_data(
+}
+#endif
+
+void Logger::set_runtime_stats() {
```
To be clear, based on the comment in the code, this sets the runtime stats of the _previous_ iteration right?

zhaojuanmao replied:
yes, let me add the comment

rohan-varma commented on torch/lib/c10d/logger.cpp:
```diff
@@ -85,6 +87,101 @@ void Logger::set_construction_logging_data(
+    calculate_avg_gpu_time(
+        ddp_logging_data_->avg_backward_compute_comm_overlap_time,
+        reducer_->gpu_timer_.backward_comm_start,
+        reducer_->gpu_timer_.backward_compute_end);
```
the ratio avg_backward_compute_comm_overlap_time/avg_backward_comm_time should always be < 1 and the better overlap we have as this approaches 1 right? Might be useful to expose this ratio in the future for interesting analysis.

zhaojuanmao replied:
you are right. And this ratio can be calculated in logging table easily. here we are trying to log raw data stats so that more flexible calculation can be done in downstream table.

rohan-varma commented on torch/lib/c10d/logger.cpp:
```diff
@@ -85,6 +87,101 @@ void Logger::set_construction_logging_data(
+        reducer_->cpu_timer_.backward_comm_start_time,
+        reducer_->cpu_timer_.backward_comm_end_time);
+
+    calculate_avg_cpu_time(
```
I didn't know comm/computation overlap was possible in CPU use cases, is this true?

zhaojuanmao replied:
In CPU cases, we can still overlap comm and computation. Bucketing helps overlap, it is not related to backend.

rohan-varma commented on torch/lib/c10d/logger.cpp:
```diff
@@ -85,6 +87,101 @@ void Logger::set_construction_logging_data(
+    cudaEventSynchronize(reducer_->gpu_timer_.backward_compute_start);
+    cudaEventSynchronize(reducer_->gpu_timer_.backward_compute_end);
+    cudaEventSynchronize(reducer_->gpu_timer_.backward_comm_start);
+    cudaEventSynchronize(reducer_->gpu_timer_.backward_comm_end);
```
If I'm reading CUDA API instructions correctly: https://developer.nvidia.com/blog/how-implement-performance-metrics-cuda-cc/ you might only need to synchronize the logically last event (last event recorded) and the rest should be implicitly synchronized?

zhaojuanmao replied:
not sure about this example, but events here do not have complete dependencies. e.g., some events may not be recorded in some no-sync iterations or when users do not call backward().... 

I added more checks and comments here before synchronizing

rohan-varma commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -3271,11 +3273,26 @@ def parse_env(var):
+            # It is hard to test accurate latency, but it can test whether the latency is
+            # a valid value and in the expected range.
+            self.assertGreaterEqual(ddp_logging_data.avg_forward_compute_time, 10)
+            self.assertGreaterEqual(ddp_logging_data.avg_backward_compute_comm_overlap_time, 10)
```
Is this unit in ns? If so, how are we sure overalp time is >= 10ns, seems like this could be flaky?

zhaojuanmao replied:
in ns, we can change this value to 1...

rohan-varma commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -3271,11 +3273,26 @@ def parse_env(var):
             self.assertEqual(ddp_logging_data.nccl_nthreads, parse_env("NCCL_NTHREADS"))
             self.assertEqual(ddp_logging_data.nccl_ib_timeout, parse_env("NCCL_IB_TIMEOUT"))
+            # test runtime logging fields
+            self.assertEqual(ddp_logging_data.unused_parameter_size, 0)
```
Can we add a test where there are unused paramters and verify this value (can be done in a follow up diff)?

rohan-varma reviewed (commented):
Looks good overall, but left a bunch of comments for your consideration. Thanks!

zhaojuanmao committed:
Update on "add stats that can only be collected at runtime"


add stats such as rebuilt bucket stats, unused parameter stats and performance stats to ddp logging data

Differential Revision: [D26158645](https://our.internmc.facebook.com/intern/diff/D26158645/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "add stats that can only be collected at runtime"


add stats such as rebuilt bucket stats, unused parameter stats and performance stats to ddp logging data

Differential Revision: [D26158645](https://our.internmc.facebook.com/intern/diff/D26158645/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "add stats that can only be collected at runtime"


add stats such as rebuilt bucket stats, unused parameter stats and performance stats to ddp logging data

Differential Revision: [D26158645](https://our.internmc.facebook.com/intern/diff/D26158645/)

[ghstack-poisoned]

pritamdamania87 commented on torch/nn/parallel/distributed.py:
```diff
@@ -673,6 +673,9 @@ def no_sync(self):
 
     def forward(self, *inputs, **kwargs):
+        if torch.is_grad_enabled() and self.require_backward_grad_sync:
+            self.logger.set_runtime_stats()
```
Have we profiled this? It might be expensive to compute stats for every iteration. We might want to do some sampling here.

zhaojuanmao replied:
yes, I profiled. There is no regression for both resnet and bert models. I think the main reason is: 1) event is initialized only once (event initialization is expensive). 2) event synchronization happens after optimizer step is done. so synchronization is cheap here. 3) event recording and average operation is actually cheap here. 

but let me think about whether we want to add sampling for safety and flexibility.

zhaojuanmao replied:
by the way, although we collect stats at every iteration, we only log data to downstream stable at 3 selected iterations. so the time spending on logging to downstream table is also trivial

pritamdamania87 commented on torch/lib/c10d/logger.hpp:
```diff
@@ -21,8 +21,29 @@ class Logger {
+  // to record the run time stats of iterations previouly ran.
+  // GPU performance stats are collected only for single process
+  // single device program right now.
+  // TODO to support single process multiple devices, events need
+  // to be created and recorded on multiple devices.
```
Do we need to do this since the plan is to deprecate single process multiple device mode?

zhaojuanmao replied:
may be not. will remove this after we finally deprecate single process multiple device mode in DDP...

pritamdamania87 commented on torch/lib/c10d/reducer.cpp:
```diff
@@ -566,7 +567,7 @@ void Reducer::mark_variable_ready(VariableIndex index) {
       "Out of range variable index.");
   backward_stats_[replica_index][variable_index] =
-      current_time_in_nanos() - backward_stats_base_;
+      current_time_in_nanos() - cpu_timer_.backward_compute_start_time;
```
Do we need nanosecond level precision? The autograd profiler reports times in microseconds which should be sufficient?

zhaojuanmao replied:
it should be no harm to use nanosecond, especially for small program that runs very fast

zhaojuanmao committed:
Update on "add stats that can only be collected at runtime"


add stats such as rebuilt bucket stats, unused parameter stats and performance stats to ddp logging data

1. gpu time stats are not collected for single process multiple devices in this diff, as that requires events are created and recorded on multiple devices
2. use at::cuda::event API for safer calls
3. events may not be created in autograd hook if hook is not triggered in user's codes, e.g., users runs in non-sync mode in some iterations. So we checked events are created or not before synchronizing, also skipped invalid results.
4. users may not set device upfront, so explicitly set proper device before creating events in our prepare_forward() and prepare_backward() calls

Differential Revision: [D26158645](https://our.internmc.facebook.com/intern/diff/D26158645/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "add stats that can only be collected at runtime"


add stats such as rebuilt bucket stats, unused parameter stats and performance stats to ddp logging data

1. gpu time stats are not collected for single process multiple devices and multi device modules in this diff, as that requires events are created and recorded on multiple devices
2. use at::cuda::event API for safer calls
3. events may not be created in autograd hook if hook is not triggered in user's codes, e.g., users runs in non-sync mode in some iterations. So we checked events are created or not before synchronizing, also skipped invalid results.
4. users may not set device upfront, so explicitly set proper device before creating events in our prepare_forward() and prepare_backward() calls

Differential Revision: [D26158645](https://our.internmc.facebook.com/intern/diff/D26158645/)

[ghstack-poisoned]

rohan-varma commented on torch/lib/c10d/reducer.cpp:
```diff
@@ -53,6 +54,21 @@ Reducer::Reducer(
   TORCH_CHECK(replicas_[0].size() >= 1, "Expected at least one parameter.");
 
+  // Check whether the module is multi_device_module
+  {
```
nit: do we need this extra pair of brackets?

zhaojuanmao replied:
following up code style in this constructor:)

rohan-varma commented on torch/lib/c10d/logger.cpp:
```diff
@@ -85,7 +91,145 @@ void Logger::set_construction_data_and_log(
+#ifdef USE_CUDA
+    // Cuda time stats are only collected for single process single
+    // device and single device module.
+    if (reducer_->replicas_.size() > 1 || reducer_->is_multi_device_module_) {
```
I'm assuming that multi device module is what was causing the cuda deadlock issues you mentioned? Do we know if we want to eventually support this, or is there some blocker? Could we file issues as needed?

zhaojuanmao replied:
cuda event, record and synchronize needs to be on the same device. currently we created events on a fixed device only once during whole training loop (because event creation is expensive, we should avoid creating it every iteration), single process multiple device or multiple device module will result in "invalid resource handle" cuda error when devices are not consistent with creation device.

mrshenli replied:
(don't need to add this in this PR, just for your reference) I created an example of using `CUDAEvent` to measure multi-device module here: https://gist.github.com/mrshenli/8c39c35612218e0d6d772910bca2f737#file-event_demo-py-L57-L81

The high-level idea is to first build dependencies between streams and then create the `CUDAEvent`.

zhaojuanmao replied:
thanks for the example, will follow up with you offline. we can add multi-device module support in the future if possible

rohan-varma commented on torch/lib/c10d/logger.cpp:
```diff
@@ -85,7 +91,145 @@ void Logger::set_construction_data_and_log(
+
+  if (reducer_->replicas_[0][0].is_cuda()) {
+#ifdef USE_CUDA
+    // Cuda time stats are only collected for single process single
```
Nit: Might want to log/warn once when the timing stats can't be collected so users are aware.

zhaojuanmao replied:
sounds good

rohan-varma commented on torch/lib/c10d/logger.cpp:
```diff
@@ -85,7 +91,145 @@ void Logger::set_construction_data_and_log(
+}
+
 c10::DDPLoggingData Logger::get_ddp_logging_data() {
+  std::lock_guard<std::mutex> lock(reducer_->mutex_);
```
Why does getting ddp_logging_data require reducer's mutex?

zhaojuanmao replied:
to sync up with threads that are mutating ddp_logging_data

rohan-varma commented on torch/lib/c10d/logger.cpp:
```diff
@@ -85,7 +91,145 @@ void Logger::set_construction_data_and_log(
+
+void Logger::set_runtime_stats() {
+  // Sync with reducer's data
+  std::lock_guard<std::mutex> lock(reducer_->mutex_);
```
Not sure if it's the best pattern to rely on another class's mutex to access it's private attributes, but I suppose it's probably fine since it's a friend class.

zhaojuanmao replied:
that is also what I thought, logger depends on reducer's data, they needs to be synced up.

rohan-varma commented on torch/lib/c10d/logger.cpp:
```diff
@@ -85,7 +91,145 @@ void Logger::set_construction_data_and_log(
+  // If unused_parameters_ is not empty, calculate its sizes.
+  // unused_parameters_ is calculated in forward call of
+  // each iteration.
+  for (const auto& unused_index : reducer_->unused_parameters_) {
```
This might trigger a nontrivial perf hit if significant (1M+ for example) parts of the model (maybe in something like multi-headed attention) aren't used in certain iterations.

zhaojuanmao replied:
be noted here the unused_parameters_ are parameter tensors, the number of parameter tensors is usually small. e.g., for 1G bert model, totally it has 150 parameter tensors.

rohan-varma commented on torch/lib/c10d/reducer.cpp:
```diff
@@ -969,6 +1000,24 @@ void Reducer::populate_bucket_views_out(
+    // and single device module.
+    if (replicas_.size() == 1 && !is_multi_device_module_) {
+      // Create and record event on the replicas_[0][0].device().
+      at::cuda::set_device(replicas_[0][0].device().index());
```
Nit: I believe the general pattern is to use a `DeviceGuard`.

zhaojuanmao replied:
Actually this is the root cause where deadlock happens for single process single device. when users are not explicitly set service in their python codes, the current device index will always be 0 for all ranks. In this case, rank 1 will try to create event on device 0, but rank 0 has launched nccl all reduce on device 0 ==> this will cause deadlock. 

DeviceGuard will not work here as well. It may be one of two reasons: 1) cudaEvent record internally used cudaGuard, which may be conflict with DeviceGuard outside. 2) DeviceGuard constructor will reset device to active device during DeviceGuard construction, which may cause issue (I am not sure). 

So I gave up using DeviceGuard. On the other hand, I think set_device() call should be safe here, as we only set_device() for single process single device program that usually requires users to explicitly set device on python upfront as well.

rohan-varma replied:
Nice catch! Sounds like the deadlock issue is quite similar to what we discussed offline with the deadlock in the profiler as well.

zhaojuanmao commented:
although current diff did not result in perf regression in our benchmarks, but for better control performance just in case some application needs it in the future, I will add sampling for event recording in this diff.

zhaojuanmao commented:
@ilia-cher and @ngimel, would you please kindly help reviewing whether CUDA time measurement parts are written properly (any pitfall I did not realize)? thanks a lot!

rohan-varma reviewed (approved):
Looks good from my side! Thank you for working through all of the different timing issues!

zhaojuanmao committed:
Update on "add stats that can only be collected at runtime"


add stats such as rebuilt bucket stats, unused parameter stats and performance stats to ddp logging data

1. gpu time stats are not collected for single process multiple devices and multi device modules in this diff, as that requires events are created and recorded on multiple devices
2. use at::cuda::event API for safer calls
3. events may not be created in autograd hook if hook is not triggered in user's codes, e.g., users runs in non-sync mode in some iterations. So we checked events are created or not before synchronizing, also skipped invalid results.
4. users may not set device upfront, so explicitly set proper device before creating events in our prepare_forward() and prepare_backward() calls

Differential Revision: [D26158645](https://our.internmc.facebook.com/intern/diff/D26158645/)

[ghstack-poisoned]

zhaojuanmao commented:
added sampling control for event recording

rohan-varma commented on torch/nn/parallel/distributed.py:
```diff
@@ -676,6 +676,9 @@ def no_sync(self):
 
     def forward(self, *inputs, **kwargs):
+        if torch.is_grad_enabled() and self.require_backward_grad_sync:
+            self.logger.set_runtime_stats()
```
Now that we have sampling I wonder if it's more performant to check the sampling here and only call into C++ if needed, might be faster than dropping/reacquiring GIL

zhaojuanmao replied:
in this case, we need to expose should_collect_runtime_stats() to python

rohan-varma commented on torch/lib/c10d/logger.cpp:
```diff
@@ -85,7 +91,148 @@ void Logger::set_construction_data_and_log(
+void Logger::set_runtime_stats() {
+  // Sync with reducer's data
+  std::lock_guard<std::mutex> lock(reducer_->mutex_);
+  // set runtime stats after 1st iteration is complete.
```
Update this comment?

rohan-varma commented on torch/lib/c10d/reducer.cpp:
```diff
@@ -1469,6 +1515,90 @@ void Reducer::ensure_prior_reduction_finished() {
 }
 
+bool Reducer::should_collect_runtime_stats() {
+  if (num_iterations_ > 0 && num_iterations_ % kDDPRuntimeLoggingSampleRate == 0) {
```
this means that the first time we will collect runtime stats is iteration 100? Should we make an exception and collect them at iteration 1 as well, so users can view some stats before iteration 100 (maybe for debugging purposes)?

zhaojuanmao replied:
sounds good

rohan-varma commented on torch/lib/c10d/logger.hpp:
```diff
@@ -21,14 +21,37 @@ class Logger {
+  // single device program right now.
+  // TODO to support single process multiple devices, events need
+  // to be created and recorded on multiple devices.
+  void set_runtime_stats();
```
Nit: document that this is now sampled? 

Also, if `set_runtime_stats` is now sampled, then what happens if I call `model_ddp.logger.get_logging_data()` and view the runtime stats, on, say iteration 30? Will the runtime stats correspond to the most recently sampled iteration (i.e.  first iteration in this case)? Is there any way to indicate to the user that they are not necessarily getting most recent data but rather sampled?

zhaojuanmao replied:
right, runtime stats will correspond to the most recently sampled iteration. let me clarify it in the comment...

zhaojuanmao committed:
Update on "add stats that can only be collected at runtime"


add stats such as rebuilt bucket stats, unused parameter stats and performance stats to ddp logging data

1. gpu time stats are not collected for single process multiple devices and multi device modules in this diff, as that requires events are created and recorded on multiple devices
2. use at::cuda::event API for safer calls
3. events may not be created in autograd hook if hook is not triggered in user's codes, e.g., users runs in non-sync mode in some iterations. So we checked events are created or not before synchronizing, also skipped invalid results.
4. users may not set device upfront, so explicitly set proper device before creating events in our prepare_forward() and prepare_backward() calls

Differential Revision: [D26158645](https://our.internmc.facebook.com/intern/diff/D26158645/)

[ghstack-poisoned]

mrshenli commented on torch/lib/c10d/logger.cpp:
```diff
@@ -85,7 +91,148 @@ void Logger::set_construction_data_and_log(
+      return;
+    }
+    // Check events on the replicas_[0][0].device().
+    at::cuda::set_device(reducer_->replicas_[0][0].device().index());
```
will this change the current device beyond this scope? Or should we use a device guard instead?

zhaojuanmao replied:
yes, that is also what I am concerned. I tried to use deviceGuard(), it hanged!  But yesterday I thought about it again, it hanged possibly because I did not set_device before synchronizing in the logger. 

so I tried to use deviceGuard again before recording and synchronizing, now it works fine!

mrshenli commented on torch/lib/c10d/logger.cpp:
```diff
@@ -85,7 +91,148 @@ void Logger::set_construction_data_and_log(
+  if (milliseconds < 0) {
+    return;
+  }
+  avg_time = (int(milliseconds * kMilliSecondToNanosSecond) +
```
nit: curious, will it be more efficient and accurate if we store the total time, and then calculate the `avg_time` on demand when needed?

zhaojuanmao replied:
I guess the efficiency improvement here will be very small, calculating avg_time at sampling iterations should be cheap?

zhaojuanmao commented on torch/lib/c10d/logger.cpp:
```diff
@@ -85,7 +91,148 @@ void Logger::set_construction_data_and_log(
+  if (cpu_end_time < cpu_start_time) {
+    return;
+  }
+  avg_time = ((cpu_end_time - cpu_start_time) +
+              avg_time * (num_iterations_stats_recorded_ - 1)) /
+      num_iterations_stats_recorded_;
```
yes

zhaojuanmao committed:
Update on "add stats that can only be collected at runtime"


add stats such as rebuilt bucket stats, unused parameter stats and performance stats to ddp logging data

1. gpu time stats are not collected for single process multiple devices and multi device modules in this diff, as that requires events are created and recorded on multiple devices
2. use at::cuda::event API for safer calls
3. events may not be created in autograd hook if hook is not triggered in user's codes, e.g., users runs in non-sync mode in some iterations. So we checked events are created or not before synchronizing, also skipped invalid results.
4. users may not set device upfront, so explicitly set proper device before creating events in our prepare_forward() and prepare_backward() calls

Differential Revision: [D26158645](https://our.internmc.facebook.com/intern/diff/D26158645/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "add stats that can only be collected at runtime"


add stats such as rebuilt bucket stats, unused parameter stats and performance stats to ddp logging data

1. gpu time stats are not collected for single process multiple devices and multi device modules in this diff, as that requires events are created and recorded on multiple devices
2. use at::cuda::event API for safer calls
3. events may not be created in autograd hook if hook is not triggered in user's codes, e.g., users runs in non-sync mode in some iterations. So we checked events are created or not before synchronizing, also skipped invalid results.
4. users may not set device upfront, so explicitly set proper device before creating events in our prepare_forward() and prepare_backward() calls

Differential Revision: [D26158645](https://our.internmc.facebook.com/intern/diff/D26158645/)

[ghstack-poisoned]
